User-guide: Add BMO live-iso (#284)
metal3-io/metal3-docs: docs/user-guide/src/bmo/live-iso.md

# Live-ISO

> The live-iso API in Metal3 allows to boot BaremetalHost with live-iso image instead of using IPA deploy ramdisk.

Reviewer: nit: BareMetalHost

> The live-iso API in Metal3 allows to boot BaremetalHost with live-iso image instead of using IPA deploy ramdisk.

Reviewer: When used in a normal text, it should be "live ISO" or "live CD", not "live-iso". So for example, "with a live ISO image.."

Author: Fixed in normal text usage.

> The live-iso API in Metal3 allows to boot BaremetalHost with live-iso image instead of using IPA deploy ramdisk.

Reviewer: ```suggestion
The live-iso API in Metal3 allows booting a BareMetalHost with a live ISO image instead of writing an image to the local disk using the IPA deploy ramdisk.
```

Why we need it?

In some circumstances, i.e to reduce boot time for ephemeral workloads, it may be possible to boot an iso
and not deploy any image to disk (saving the time to write the image and reboot). Since this alternative

> is distributed as a live-iso, leveraging the existing toolchains like [fedora-coreos installer](https://docs.fedoraproject.org/en-US/fedora-coreos/bare-metal/) might be desirable.

Reviewer: This sentence reads a bit awkward to me. "this alternative" refers to the live-iso mode, so it cannot be distributed. You probably wanted something like:

This API is also useful for integration with 3rd party installers distributed as a CD image, for example ...

Author: Done

How to use it?

Here is an example with a BaremetalHost CRD, where iso referenced by the url and `live-iso` set in DiskFormat will be
live-booted without deploying an image to disk. Checksum options are not required in this case, and will be ignored if specified:

> 

Reviewer: An important bit is missing: the live ISO mode is only supported for virtual media drivers.

Author: Good point, added now.

```yaml
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: live-iso-booted-node
spec:
  image:
    url: http://1.2.3.4/image.iso
    format: live-iso
  online: true
```

**Note**  `rootDeviceHints` and `userData` will not be used
since the image won't be written to disk, and Ironic doesn't

> support passing user-data in addition to the iso attachment at present.

Reviewer: It's not 100% correct, Ironic supports attaching it as a virtual USB stick. The problem is, a lot of hardware does not allow 2 virtual media devices at the same time. I'd probably skip the reasoning. But do mention that network data is also not used.

Author: I have removed the reasoning part for ironic, however did leave that `rootDeviceHints`, `networkData` and `userData` won't be used due to it image is not written to the disk. Hope that makes sense 🙂

# Live ISO

The live-iso API in Metal3 allows booting a BareMetalHost with a live ISO image instead of writing an image to
the local disk using the IPA deploy ramdisk.

Why we need it?

In some circumstances, i.e to reduce boot time for ephemeral workloads, it may be possible to boot an iso
and not deploy any image to disk (saving the time to write the image and reboot). This API is also useful
for integration with 3rd party installers distributed as a CD image, for example leveraging the existing
toolchains like [fedora-coreos installer](https://docs.fedoraproject.org/en-US/fedora-coreos/bare-metal/)
might be desirable.

How to use it?

Here is an example with a BaremetalHost CRD, where iso referenced by the url and `live-iso` set in DiskFormat will be
live-booted without deploying an image to disk. It is important to note that, live ISO mode is only supported when

> redfish-virtualmedia driver is used as BMC driver. Also, checksum options are not required in this case, and will be

Reviewer: Any virtualmedia driver actually.

Author: Thanks, fixed now

# Live-ISO

The live-iso API in Metal3 allows to boot BaremetalHost with live-iso image instead of using IPA deploy ramdisk.

Why we need it?

In some circumstances, i.e to reduce boot time for ephemeral workloads, it may be possible to boot an iso
and not deploy any image to disk (saving the time to write the image and reboot). Since this alternative

> is distributed as a live-iso, leveraging the existing toolchains like [fedora-coreos installer](https://docs.fedoraproject.org/en-US/fedora-coreos/bare-metal/) might be desirable. On top that, Ironic already exposes an [API](https://docs.openstack.org/ironic/latest/admin/ramdisk-boot.html) to support this.

Reviewer 2: I would skip the last sentence here since it is already obvious that Ironic must support it if BMO does. Or we could say something like "this is implemented using the ramdisk-boot API in Ironic" so the user can read more about it if interested.
```suggestion
is distributed as a live-iso, leveraging the existing toolchains like [fedora-coreos installer](https://docs.fedoraproject.org/en-US/fedora-coreos/bare-metal/) might be desirable.
```

Author: I got rid of it, thanks

How to use it?

Here is an example with a BaremetalHost CRD, where DiskFormat spec indicates that instead
of deploying an image to disk, a live-iso image will be booted:

```yaml
spec:

> image:
>     url: http://1.2.3.4/image.iso
>     format: live-iso
> online: true

Reviewer 2: ```suggestion
  image:
    url: http://1.2.3.4/image.iso
    format: live-iso
  online: true
```

```

**Note**  `rootDeviceHints` and `userData` will not be used
since the image won't be written to disk, and Ironic doesn't
support passing user-data in addition to the iso attachment at present.

There are two requirement to enable this feature on Ironic side:

1. Please make sure that Ironic is configured to use the [ramdisk
   deploy interface](https://docs.openstack.org/ironic/latest/admin/ramdisk-boot.html)

```yaml
enabled_deploy_interfaces = direct,ramdisk
```

2. Also, you need to set the provided ISO URL into the

>    [instance_info boot_iso field](https://docs.openstack.org/ironic/latest/admin/drivers/redfish.html#virtual-media-ramdisk)

Reviewer 2: If I understand correctly, this is what BMO will do for you behind the scenes, not something that the user has to do, so I would skip it. Or alternatively reword it so that it is clear that this is something BMO does and you don't have to think about it.
